
### What makes this tool different from other seed-free tools:
+ MEANGS does not filter low sequencing depth reads.
+ Two pipelines are available to extract and assemble either protein-coding genes (PCGs) or a complete mitogenome. The default mode was used for only PCG fragments assembly, whereas the complete mitogenome was generated using the deepin mode.
+ It uses Hidden Markov Model **(HMM)** and a Short Sequence Assembly by progressive K-mer search and 3′ read Extension **(SSAKE)** algorithm.

## Step-by-step

MEANGS uses a HMM database that contains mitogenome sequence conservation modelling information from different taxo- nomic clades in Metazoans from MitoZ.

### 1. Seed generation

+ First, MEANGS using seqtk filters the raw data generated by High-Throughput Screening (HTS) platforms by removing reads characterized by low quality (q≤10), short read length (≤ 30 bp) or high error rate (q≥0.05) and was used to convert the reads into fasta formatted sequences.
+ Then, the program uses profile HMMs to select candidate mitochondrial reads. Reads with a high confidence score S were picked and assembled to generate candidate mitochondrial PCG fragments.
+ The assembled fragments are filter through HMMs again to avoid "redundant fragments".

### NOTE 
> What that fuck is a Hidden Markov Model and how does it work?
> [[HMM incompleto]]


